Libraries/LibCore/Account.cpp

```diff
@@ -0,0 +1,202 @@
+
+    return builder.build();
+}
+
+static Vector<gid_t> get_gids(const StringView& username)
+{
+    Vector<gid_t> extra_gids;
+    for (auto* group = getgrent(); group; group = getgrent()) {
+        for (size_t i = 0; group->gr_mem[i]; ++i) {
+            if (username == group->gr_mem[i]) {
+                extra_gids.append(group->gr_gid);
+                break;
+            }
+        }
+    }
+    endgrent();
+    return extra_gids;
+}
+
+Result<Account, String> Account::from_name(const char* username)
+{
+    struct passwd* pwd = nullptr;
+    errno = 0;
+    pwd = getpwnam(username);
+    if (!pwd) {
+        if (errno == 0)
+            return String("No such user");
+
+        return String(strerror(errno));
+    }
+
+    Account account(pwd, get_gids(pwd->pw_name));
+    endpwent();
+    return account;
+}
+
+Result<Account, String> Account::from_uid(uid_t uid)
+{
+    struct passwd* pwd = nullptr;
+    errno = 0;
+    pwd = getpwuid(uid);
+    if (!pwd) {
+        if (errno == 0)
+            return String("No such user");
+
+        return String(strerror(errno));
+    }
+
+    Account account(pwd, get_gids(pwd->pw_name));
+    endpwent();
+    return account;
+}
+
+bool Account::authenticate(const char* password) const
+{
+    // An empty passwd field indicates that no password is required to log in.
+    if (m_password_hash.is_empty())
+        return true;
+
+    // FIXME: Use crypt_r if it can be built in lagom.
+    char* hash = crypt(password, m_password_hash.characters());
+    return hash != nullptr && strcmp(hash, m_password_hash.characters()) == 0;
+}
+
+bool Account::login() const
+{
+    if (setgroups(m_extra_gids.size(), m_extra_gids.data()) < 0)
+        return false;
+
+    if (setgid(m_gid) < 0)
+        return false;
+
+    if (setuid(m_uid) < 0)
+        return false;
+
+    return true;
+}
+
+void Account::set_password(const char* password)
+{
+    m_password_hash = crypt(password, get_salt().characters());
+}
+
+void Account::set_password_enabled(bool enabled)
+{
+    if (enabled && m_password_hash != "" && m_password_hash[0] == '!') {
+        m_password_hash = m_password_hash.substring(1, m_password_hash.length() - 1);
+    } else if (!enabled && (m_password_hash == "" || m_password_hash[0] != '!')) {
+        StringBuilder builder;
+        builder.append('!');
+        builder.append(m_password_hash);
+        m_password_hash = builder.build();
+    }
+}
+
+void Account::delete_password()
+{
+    m_password_hash = "";
+}
+
+Account::Account(struct passwd* pwd, Vector<gid_t> extra_gids)
+    : m_username(pwd->pw_name)
+    , m_password_hash(pwd->pw_passwd)
+    , m_uid(pwd->pw_uid)
+    , m_gid(pwd->pw_gid)
+    , m_gecos(pwd->pw_gecos)
+    , m_home_directory(pwd->pw_dir)
+    , m_shell(pwd->pw_shell)
+    , m_extra_gids(extra_gids)
+{
+}
+
+bool Account::sync()
+{
+    StringBuilder new_passwd_file;
+
+    setpwent();
+
+    struct passwd* p;
+    while ((p = getpwent())) {
```

Comment:
This, too, can fail. If we don't distinguish errors from reaching the end of users, we'll just end up truncating the file, and so silently dropping some users.